### PR TITLE
Add support for named function type arguments

### LINKED ISSUE #180: Support named type arguments
https://github.com/Roblox/luau/blob/master/rfcs/syntax-named-function-type-args.md
https://devforum.roblox.com/t/luau-recap-may-2021/1268364

```lua
type MyCallbackType = (cost: number, name: string) -> string

local cb: (amount: number) -> number
local function foo(cb: (name: string) -> ())
```

### PR DESCRIPTION
**Pull request by JohnnyMorganz**
Closes #180 

This PR may have had a bit of scope creep, as I completely redid how parsing parentheses works (again). This time, I've split them all up into smaller parsing blocks, to hopefully make them easier to reason about.

The parsing now better represents how Luau parsing is done in Roblox Studio [I think..., this is all really just done by testing and seeing what errors pop up and what they say]

### CODE REVIEW

**Comment by Kampfkarren** [comment]
Conflicts, your "aa" folder is back.

File: full-moon/src/ast/parsers.rs [hunk 1]
```diff
@@ -1374,66 +1465,48 @@ cfg_if::cfg_if! {
...
-                    (
-                        state,
-                        TypeInfo::Callback {
-                            arguments: types,
-                            parentheses: ContainedSpan::new(start_parenthese, end_parenthese),
-                            arrow,
-                            return_type: Box::new(return_value),
-                        },
-                    )
+                // TODO: can we get rid of this clone?
```
Comment (Kampfkarren) [inline full-moon/src/ast/parsers.rs:1468]: Same question

File: full-moon/src/ast/parsers.rs [hunk 2]
```diff
@@ -1374,66 +1465,48 @@ cfg_if::cfg_if! {
...
+                            TypeInfo::Tuple { parentheses, types } => {
+                                // `types` is currently `Punctuated<TypeInfo>`, but we need to map it to `Punctuated<TypeArgument>`
+                                // where each argument has no name.
+                                let mut arguments = Punctuated::new();
+                                for pair in types.into_pairs() {
+                                    arguments.push(pair.map(|type_info| TypeArgument {
+                                        name: None,
+                                        type_info
+                                    }))
+                                };
```
Comment (Kampfkarren) [inline full-moon/src/ast/parsers.rs:1484]: `Punctuated::from_iter(types.into_pairs().map(...))`

File: full-moon/src/ast/parsers.rs [hunk 3]
```diff
@@ -1496,6 +1569,82 @@ cfg_if::cfg_if! {
...
+                    "expected `)` to match `(`"
+                );
+
+                // Single token lookahead: if we see a `->` afterwards, we need to bail out
+                // of attempting to parse a tuple type, since it must be a callback.
+                if let Ok((state, arrow)) = ParseSymbol(Symbol::ThinArrow).parse(state) {
+                    // TODO: this is copied from converting parentheses -> callback above
+                    // `types` is currently `Punctuated<TypeInfo>`, but we need to map it to `Punctuated<TypeArgument>`
+                    // where each argument has no name.
+                    let mut arguments = Punctuated::new();
```
Comment (Kampfkarren) [inline full-moon/src/ast/parsers.rs:1603]: Same here, Punctuated::from_iter.

File: full-moon/src/ast/types.rs [hunk 4]
```diff
@@ -526,6 +526,56 @@ impl<'a> TypeSpecifier<'a> {
...
+    /// Creates a new TypeArgument with the given type info
+    pub fn new(type_info: TypeInfo<'a>) -> Self {
+        Self {
+            name: None,
+            type_info,
+        }
+    }
+
+    /// The name of the argument split into identifier and punctuation: `count:` in `count: number`.
+    pub fn name(&self) -> Option<&(TokenReference<'a>, TokenReference<'a>)> {
```
Comment (Kampfkarren) [inline full-moon/src/ast/types.rs:549]: I forget, are our other tuple results like `Option<&(T1, T2)>`? I thought they were `Option<(&T1, &T2)>`?

Comment (JohnnyMorganz) [reply full-moon/src/ast/types.rs:549]: In Block it seems to be `Option<&(T1, T2)>`

Comment (Kampfkarren) [reply full-moon/src/ast/types.rs:549]: Alright, works for me.

**Pull request merged**